ydb/core/kqp/provider/rewrite_io_utils.cpp

```diff
@@ -0,0 +1,169 @@
+#include "rewrite_io_utils.h"
+
+#include <ydb/library/yql/core/expr_nodes/yql_expr_nodes.h>
+#include <ydb/library/yql/core/yql_expr_optimize.h>
+#include <ydb/library/yql/providers/common/provider/yql_provider.h>
+#include <ydb/library/yql/providers/common/provider/yql_provider_names.h>
+#include <ydb/library/yql/sql/sql.h>
+#include <ydb/library/yql/utils/log/log.h>
+
+namespace NYql {
+namespace {
+
+using namespace NNodes;
+
+constexpr const char* QueryGraphNodeSignature = "SavedQueryGraph";
+
+NSQLTranslation::TTranslationSettings CreateViewTranslationSettings(const TString& cluster) {
+    NSQLTranslation::TTranslationSettings settings;
+
+    settings.DefaultCluster = cluster;
+    settings.ClusterMapping[cluster] = TString(NYql::KikimrProviderName);
+    settings.Mode = NSQLTranslation::ESqlMode::LIMITED_VIEW;
+
+    return settings;
+}
+
+TExprNode::TPtr CompileQuery(
+    const TString& query,
+    TExprContext& ctx,
+    const TString& cluster
+) {
+    TAstParseResult queryAst;
+    queryAst = NSQLTranslation::SqlToYql(query, CreateViewTranslationSettings(cluster));
+
+    ctx.IssueManager.AddIssues(queryAst.Issues);
+    if (!queryAst.IsOk()) {
+        return nullptr;
+    }
+
+    TExprNode::TPtr queryGraph;
+    if (!CompileExpr(*queryAst.Root, queryGraph, ctx, nullptr, nullptr)) {
+        return nullptr;
+    }
+
+    return queryGraph;
+}
+
+bool ContainsNullNode(const TExprNode::TPtr& root) {
+    return static_cast<bool>(FindNode(root, [](const TExprNode::TPtr& node) {
+        return !node;
+    }));
+}
+
+void AddChild(const TExprNode::TPtr& parent, const TExprNode::TPtr& newChild) {
+    auto childrenToChange = parent->ChildrenList();
+    childrenToChange.emplace_back(newChild);
+    parent->ChangeChildrenInplace(std::move(childrenToChange));
+}
+
+void ChangeChild(const TExprNode::TPtr& parent, ui32 index, const TExprNode::TPtr& newChild) {
+    YQL_ENSURE(parent->ChildrenSize() > index);
+
+    auto childrenToChange = parent->ChildrenList();
+    childrenToChange[index] = newChild;
+    parent->ChangeChildrenInplace(std::move(childrenToChange));
+}
+
+TExprNode::TPtr FindSavedQueryGraph(const TExprNode::TPtr& carrier) {
+    if (carrier->ChildrenSize() == 0) {
+        return nullptr;
+    }
+    auto lastChild = carrier->Children().back();
+    return lastChild->IsCallable(QueryGraphNodeSignature) ? lastChild->ChildPtr(0) : TExprNode::TPtr();
+}
+
+void SaveQueryGraph(const TExprNode::TPtr& carrier, TExprContext& ctx, const TExprNode::TPtr& payload) {
+    AddChild(carrier, ctx.NewCallable(payload->Pos(), QueryGraphNodeSignature, {payload}));
+}
+
+void InsertExecutionOrderDependencies(const TExprNode::TPtr& queryGraph, const TExprNode::TPtr& worldBefore) {
+    VisitExpr(
+        queryGraph,
+        nullptr,
+        [&worldBefore](const TExprNode::TPtr& node) {
+            if (node->ChildrenSize() > 0 && node->Child(0)->IsWorld()) {
+                ChangeChild(node, 0, worldBefore);
```

Comment:
I'm not sure changing nodes inside Visit is a good idea. We have ctx.ReplaceNodes method for this.

Reply:
I tried using both RemapExpr and OptimizeExpr and they don't seem to work for my case. I suspect the following [line](https://github.com/ydb-platform/ydb/blob/0a965144219a01dadef957ee280b78147d0667f7/ydb/library/yql/core/yql_expr_optimize.cpp#L94C1-L98C1):
```cpp
if (... || node->Type() == TExprNode::World) {
    // don't optimise node
}
```

So, I [ended up](https://github.com/ydb-platform/ydb/blob/504dad86b6f2606644cac43538b293c19ad15902/ydb/core/kqp/provider/rewrite_io_utils.cpp#L78) using ctx.ReplaceNode.